dipy/tracking/streamline.py

```diff
@@ -323,3 +327,130 @@ def orient_by_rois(streamlines, roi1, roi2, affine=None, copy=True):
             new_sl[idx] = sl[::-1]
 
     return new_sl
+
+
+def _extract_vals(data, streamlines, affine=None, threedvec=False):
+    """
+    Helper function for use with `values_from_volume`.
+
+    Parameters
+    ----------
+    data : 3D or 4D array
+        Scalar (for 3D) and vector (for 4D) values to be extracted. For 4D
+        data, interpolation will be done on the 3 spatial dimensions in each
+        volume.
+
+    streamlines : ndarray or list
+        If array, of shape (n_streamlines, n_nodes, 3)
+        If list, len(n_streamlines) with (n_nodes, 3) array in
+        each element of the list.
+
+    affine : ndarray, shape (4, 4)
+        Affine transformation from voxels (image coordinates) to streamlines.
+        Default: identity.
+
+    threedvec : bool
+        Whether the last dimension has length 3. This is a special case in
+        which we can use :func:`vfu.interpolate_vector_3d` for the
+        interploation of 4D volumes without looping over the elements of the
+        last dimension.
+
+    Return
+    ------
+    array or list (depending on the input) : values interpolate to each
+        coordinate along the length of each streamline
+    """
+    data = data.astype(np.float)
+    if affine is not None:
+        streamlines = ut.move_streamlines(streamlines,
+                                          np.linalg.inv(affine))
+
+    if (isinstance(streamlines, list) or
+            isinstance(streamlines, types.GeneratorType)):
+        vals = []
+        for sl in streamlines:
+            if threedvec:
+                vals.append(list(vfu.interpolate_vector_3d(data, sl)[0]))
```

Comment:
On Fri, Jan 15, 2016 at 7:15 AM, Samuel St-Jean <notifications@github.com>
wrote:

> In dipy/tracking/streamline.py
> <https://github.com/nipy/dipy/pull/811#discussion_r49863546>:
>
> > +    Return
> > +    ------
> > +    array or list (depending on the input) : values interpolate to each
> > +        coordinate along the length of each streamline
> > +    """
> > +    data = data.astype(np.float)
> > +    if affine is not None:
> > +        streamlines = ut.move_streamlines(streamlines,
> > +                                          np.linalg.inv(affine))
> > +
> > +    if (isinstance(streamlines, list) or
> > +            isinstance(streamlines, types.GeneratorType)):
> > +        vals = []
> > +        for sl in streamlines:
> > +            if threedvec:
> > +                vals.append(list(vfu.interpolate_vector_3d(data, sl)[0]))
>
> The interpolate function only works on float64, so it gives errors like
>
> File "/home/samuel/python/dipy/tracking/streamline.py", line 440, in
> values_from_volume
> affine=affine))
> File "/home/samuel/python/dipy/tracking/streamline.py", line 375, in
> _extract_vals
> vals.append(list(vfu.interpolate_scalar_3d(data, sl)[0]))
> File "dipy/align/vector_fields.pyx", line 434, in
> dipy.align.vector_fields.interpolate_scalar_3d
> (dipy/align/vector_fields.c:21841)
> ValueError: Buffer dtype mismatch, expected 'double' but got 'float'
>
> just adding a quick n dirty sl=sl.astype(np.float64) makes it work though,
> so at this point it might be worth to do the typecast in the cython files
> directly instead of each upper functions.
>
Sorry - is this happening somewhere in our CI? Or in some test that you've
written? I am not seeing this anywhere. Could you share some code that
triggers this error with this function?



> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/811/files#r49863546>.
>